src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -389,6 +402,320 @@ private void generateNgxGraphRepresentationForRelation(CompetencyRelation relati
+        Set<NgxLearningPathDTO.Node> nodes = new HashSet<>();
+        Set<NgxLearningPathDTO.Edge> edges = new HashSet<>();
+
+        var recommendedOrderOfCompetenciesById = getRecommendedOrderOfCompetencies(learningPath);
+        var recommendedOrderOfCompetencies = recommendedOrderOfCompetenciesById.stream()
+                .map(id -> learningPath.getCompetencies().stream().filter(competency -> competency.getId().equals(id)).findFirst().get()).toList();
+
+        // generate ngx representation of recommended competencies
+        // IMPORTANT generateNgxGraphRepresentationForCompetency will be replaced by future PR
+        recommendedOrderOfCompetencies.forEach(competency -> generateNgxGraphRepresentationForCompetency(learningPath, competency, nodes, edges));
+        // generate edges between competencies
+        for (int i = 0; i < recommendedOrderOfCompetencies.size() - 1; i++) {
+            var sourceNodeId = getCompetencyEndNodeId(recommendedOrderOfCompetenciesById.get(i));
+            var targetNodeId = getCompetencyStartNodeId(recommendedOrderOfCompetenciesById.get(i + 1));
+            edges.add(new NgxLearningPathDTO.Edge(getRelationEdgeId(sourceNodeId, targetNodeId), sourceNodeId, targetNodeId));
+        }
+
+        return new NgxLearningPathDTO(nodes, edges);
+    }
+
+    /**
+     * Analyzes the current progress within the learning path and generates a recommended ordering of competencies.
+     *
+     * @param learningPath the learning path that should be analyzed
+     * @return the recommended ordering of competencies
+     */
+    private List<Long> getRecommendedOrderOfCompetencies(LearningPath learningPath) {
+        RecommendationState state = generateInitialRecommendationState(learningPath);
+        var pendingCompetencies = getPendingCompetencies(learningPath.getCompetencies(), state);
+        return simulateProgression(pendingCompetencies, state);
+    }
+
+    /**
+     * Generates the initial state of the recommendation containing all necessary information for the prediction.
+     *
+     * @param learningPath the learning path that should be analyzed
+     * @return the initial RecommendationState
+     * @see RecommendationState
+     */
+    private RecommendationState generateInitialRecommendationState(LearningPath learningPath) {
+        HashMap<Long, Set<Long>> matchingClusters = getMatchingCompetencyClusters(learningPath.getCompetencies());
+        HashMap<Long, Set<Long>> priorsCompetencies = getPriorCompetencyMapping(learningPath.getCompetencies(), matchingClusters);
+        HashMap<Long, Long> extendsCompetencies = getExtendsCompetencyMapping(learningPath.getCompetencies(), matchingClusters, priorsCompetencies);
+        HashMap<Long, Long> assumesCompetencies = getAssumesCompetencyMapping(learningPath.getCompetencies(), matchingClusters, priorsCompetencies);
+        Set<Long> masteredCompetencies = new HashSet<>();
+        HashMap<Long, Double> competencyMastery = new HashMap<>();
```

Comment:
This only stores the competencies that are not mastered yet?

Reply:
Yes